### PR TITLE
Keyvaluestore: queue transaction fast path changes

### PR DESCRIPTION
**Pull request by varadakari**
Short circuiting the queue transaction path

Signed-off-by: Varada Kari <varada.kari@sandisk.com>

### CODE REVIEW

**Comment by yuyuyu101** [comment]
@liewegas @athanatos I just think it's of do transaction inline when pg ensure the isolation. Do we have other exceptions?

@varadakari As I mentioned before, do you think move "submit_transaction_sync" to another thread could improve bandwidth? From leveldb/rocksdb backend, I would like to think it could help

**Comment by varadakari** [comment]
@yuyuyu101 do you want to have one more thread pool/queue for the sync transactions ? i mean if we want to make one more thread switch, we have to wait for the completion of commit. That would add some more latencies to the write path right? or am i missing something?

**Comment by yuyuyu101** [comment]
@varadakari yes, for one io depth workload, it may add extra context swith. But for high io depth, it could reduce lots of work at leveldb/rocksdb level I think. Lots of sync request could be batched.

**Comment by xiaoxichen** [comment]
@yuyuyu101 , are you talking about the same architecture like newstore? There is a sync thread that periodically do a sync transaction to ensure "ondisk" while keeping other transaction async?

**Comment by varadakari** [comment]
@yuyuyu101 sure, let me explore some more in that direction. One good use case can be clone.

**Comment by somnathr** [comment]
Varada,
I think we need to add a config option for the sync and async.. if we go for sync route for every back end slower backend will be a problem..

**Comment by yuyuyu101** [comment]
@somnathr because inline submit_transaction_sync will be blocked too much time for pg threads. it will reduce high io depth benchmark bandwidth. Batch write would be great for this PR change I think.

File: src/os/KeyValueStore.cc [hunk 1]
```diff
@@ -1098,23 +1092,6 @@ KeyValueStore::Op *KeyValueStore::build_op(list<Transaction*>& tls,
...
-void KeyValueStore::queue_op(OpSequencer *osr, Op *o)
-{
-  // queue op on sequencer, then queue sequencer for the threadpool,
-  // so that regardless of which order the threads pick up the
-  // sequencer, the op order will be preserved.
-
-  osr->queue(o);
-
-  perf_logger->inc(l_os_ops);
-  perf_logger->inc(l_os_bytes, o->bytes);
```
Comment (yuyuyu101) [inline src/os/KeyValueStore.cc]: we need to move these

File: src/os/KeyValueStore.cc [hunk 2]
```diff
@@ -1162,50 +1139,6 @@ void KeyValueStore::op_queue_release_throttle(Op *o)
...
-
-  utime_t lat = ceph_clock_now(g_ceph_context);
-  lat -= o->start;
-
-  dout(10) << "_finish_op " << o << " seq " << o->op << " " << *osr << "/" << osr->parent << " lat " << lat << dendl;
-  osr->apply_lock.Unlock();  // locked in _do_op
-  op_queue_release_throttle(o);
-
-  perf_logger->tinc(l_os_commit_lat, lat);
-  perf_logger->tinc(l_os_apply_lat, lat);
```
Comment (yuyuyu101) [inline src/os/KeyValueStore.cc]: ditto

**Comment by dachary** [comment]
please ignore the ubuntu bot fail, it is a false negative.

**Comment by yuyuyu101** [comment]
@varadakari plz rebase this

**Comment by varadakari** [comment]
@yuyuyu101 i am working on comments, to add a sync_thread. will rebase and open the pull request with the new set of changes.

**Pull request closed without merging**